e2e test "delete stacks with propagation=Foreground" is flaky

Executing this test may lead to a crash of the compose controller pod

```
time="2019-02-21T16:48:19Z" level=debug msg="Sending stack deletion request: e2e-tests-compose-481797610012260660/app"
time="2019-02-21T16:48:19Z" level=error msg="Unable to get stack \"e2e-tests-compose-481797610012260660/app\" owner: stacks.compose.docker.com \"app\" not found"
panic: fatal error: controller cannot retrive ownership information from a stack

goroutine 85 [running]:
github.com/docker/compose-on-kubernetes/internal/controller.(*stackOwnerCache).get(0xc0000980e0, 0xc0001fcb40, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/docker/compose-on-kubernetes/internal/controller/stackownercache.go:131 +0x175
github.com/docker/compose-on-kubernetes/internal/controller.(*impersonatingResourceUpdaterProvider).getUpdater(0xc0001bc3c0, 0xc0001fcb40, 0x0, 0xc00052c780, 0x0, 0x0, 0x28)
	/go/src/github.com/docker/compose-on-kubernetes/internal/controller/resourceupdater.go:24 +0x7f
github.com/docker/compose-on-kubernetes/internal/controller.(*StackReconciler).deleteStackChildren(0xc000682240, 0xc0001fcb40)
	/go/src/github.com/docker/compose-on-kubernetes/internal/controller/stackreconciler.go:114 +0x219
github.com/docker/compose-on-kubernetes/internal/controller.(*StackReconciler).Start.func1(0xc00048e240, 0xc00048e420, 0xc000682240, 0xc0000803c0)
	/go/src/github.com/docker/compose-on-kubernetes/internal/controller/stackreconciler.go:80 +0x19a
created by github.com/docker/compose-on-kubernetes/internal/controller.(*StackReconciler).Start
	/go/src/github.com/docker/compose-on-kubernetes/internal/controller/stackreconciler.go:70 +0x6d
```

--- a/internal/controller/resourceupdater.go
+++ b/internal/controller/resourceupdater.go
@@ -21,13 +21,15 @@ type impersonatingResourceUpdaterProvider struct {
 }
 
 func (p *impersonatingResourceUpdaterProvider) getUpdater(stack *latest.Stack, isDirty bool) (resourceUpdater, error) {
-	ic := p.ownerCache.get(stack, !isDirty)
+	ic, err := p.ownerCache.getWithRetries(stack, !isDirty)
+	if err != nil {
+		return nil, err
+	}
 	localConfig := p.config
 	localConfig.Impersonate = ic
 	result := &k8sResourceUpdater{
 		originalStack: stack,
 	}
-	var err error
 	if result.stackClient, err = clientset.NewForConfig(&localConfig); err != nil {
 		return nil, err
 	}

--- a/internal/controller/stacklistener_test.go
+++ b/internal/controller/stacklistener_test.go
@@ -31,8 +31,8 @@ func (s *dummyOwnerCache) setDirty(key string) {
 	}
 }
 
-func (s *dummyOwnerCache) get(stack *latest.Stack, acceptDirty bool) rest.ImpersonationConfig {
-	return rest.ImpersonationConfig{}
+func (s *dummyOwnerCache) getWithRetries(stack *latest.Stack, acceptDirty bool) (rest.ImpersonationConfig, error) {
+	return rest.ImpersonationConfig{}, nil
 }
 func TestStackListenerCacheInvalidation(t *testing.T) {
 	cache := &dummyOwnerCache{

--- a/internal/controller/stackownercache.go
+++ b/internal/controller/stackownercache.go
@@ -5,10 +5,12 @@ import (
 	"time"
 
 	"github.com/docker/compose-on-kubernetes/api/client/clientset"
+	"github.com/docker/compose-on-kubernetes/api/client/clientset/scheme"
 	"github.com/docker/compose-on-kubernetes/api/compose/latest"
 	"github.com/docker/compose-on-kubernetes/internal/stackresources"
 	log "github.com/sirupsen/logrus"
 	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/util/wait"
 	"k8s.io/client-go/rest"
 )
@@ -41,7 +43,12 @@ type apiOwnerGetter struct {
 
 func (g *apiOwnerGetter) get(stack *latest.Stack) (*latest.Owner, error) {
 	var owner latest.Owner
-	if err := g.Get().Namespace(stack.Namespace).Name(stack.Name).Resource("stacks").SubResource("owner").Do().Into(&owner); err != nil {
+	if err := g.Get().Namespace(stack.Namespace).Name(stack.Name).
+		Resource("stacks").
+		SubResource("owner").
+		VersionedParams(&metav1.GetOptions{IncludeUninitialized: true}, scheme.ParameterCodec).
+		Do().
+		Into(&owner); err != nil {
 		return nil, err
 	}
 	return &owner, nil

--- a/internal/controller/stackownercache.go
+++ b/internal/controller/stackownercache.go
@@ -19,7 +19,7 @@ import (
 type StackOwnerCacher interface {
 	remove(key string)
 	setDirty(key string)
-	get(stack *latest.Stack, acceptDirty bool) rest.ImpersonationConfig
+	getWithRetries(stack *latest.Stack, acceptDirty bool) (rest.ImpersonationConfig, error)
 }
 
 type ownerGetter interface {
@@ -112,9 +112,9 @@ func (s *stackOwnerCache) getWithError(stack *latest.Stack, acceptDirty bool) (r
 	return ic, nil
 }
 
-func (s *stackOwnerCache) get(stack *latest.Stack, acceptDirty bool) rest.ImpersonationConfig {
+func (s *stackOwnerCache) getWithRetries(stack *latest.Stack, acceptDirty bool) (rest.ImpersonationConfig, error) {
 	var ic rest.ImpersonationConfig
-	if err := wait.ExponentialBackoff(wait.Backoff{
+	err := wait.ExponentialBackoff(wait.Backoff{
 		Duration: time.Second,
 		Factor:   2,
 		Jitter:   float64(100 * time.Millisecond),
@@ -134,8 +134,6 @@ func (s *stackOwnerCache) get(stack *latest.Stack, acceptDirty bool) rest.Impers
 			return false, err
 		}
 		return false, nil
-	}); err != nil {
-		panic("fatal error: controller cannot retrive ownership information from a stack")
-	}
-	return ic
+	})
+	return ic, err
 }

--- a/internal/controller/stackownercache_test.go
+++ b/internal/controller/stackownercache_test.go
@@ -38,19 +38,19 @@ func TestUpdateDeleteSequence(t *testing.T) {
 	testStack.Name = "test"
 	testStack.Namespace = "ns"
 	// as of create
-	cfg := testee.get(testStack, false)
+	cfg, _ := testee.getWithRetries(testStack, false)
 	assert.Equal(t, "test", cfg.UserName)
 	// as of update
 	testee.setDirty(stackresources.ObjKey(testStack.Namespace, testStack.Name))
-	cfg = testee.get(testStack, false)
+	cfg, _ = testee.getWithRetries(testStack, false)
 	assert.Equal(t, "test", cfg.UserName)
 	// as of update followed by delete
 	testee.setDirty(stackresources.ObjKey(testStack.Namespace, testStack.Name))
 	return404 = true
-	cfg = testee.get(testStack, false)
+	cfg, _ = testee.getWithRetries(testStack, false)
 	assert.Equal(t, "test", cfg.UserName)
 	// as of delete
-	cfg = testee.get(testStack, true)
+	cfg, _ = testee.getWithRetries(testStack, true)
 	assert.Equal(t, "test", cfg.UserName)
 	testee.remove(stackresources.ObjKey(testStack.Namespace, testStack.Name))
 
@@ -70,7 +70,6 @@ func TestStackOwnerCachePanicOnUnresolvableOwner(t *testing.T) {
 	testStack := &latest.Stack{}
 	testStack.Name = "test"
 	testStack.Namespace = "ns"
-	assert.Panics(t, func() {
-		testee.get(testStack, true)
-	})
+	_, err := testee.getWithRetries(testStack, true)
+	assert.Error(t, err)
 }

